### ISSUE TITLE
spring-security-test: @WithMockOidcUser

### ISSUE THREAD

**Issue by nenaraab** [spring-projects/spring-security#8459]
**Expected Behavior**
similar to [`@WithMockUser`](https://docs.spring.io/spring-security/site/docs/current/reference/html5/#test-method-withmockuser):

```java
    @Test
    @WithMockOidcUser(name = "Any@unknown.org" )
    public void authenticateWithoutPermission_200() throws Exception {
        mockMvc.perform(get("/authenticate"))
                .andExpect(status().isOk());
    }
```

**Current Behavior**
I'm unable to use Spring WebMVC tests for my secured methods and those methods that expects the security context to be filled. They require the user id (name) from the Oidc token also to perform authorization checks with `@PreAuthorize`. 

**Context**
My application requires a `OidcUser` as AuthenticationPrincipal and uses [Method Security](https://docs.spring.io/spring-security/site/docs/current/reference/html5/#method-security-expressions).

**Comment by eleftherias** [comment]
@nenaraab This functionality is available as of Spring Security 5.3.
Check our [this section](https://docs.spring.io/spring-security/site/docs/5.3.2.RELEASE/reference/html5/#testing-oidc-login) of the reference documentation.

**Comment by rwinch** [comment]
@eleftherias This functionality is a little different given it is annotation based. We can leave this closed as a duplicate of the PR gh-8461 There is some discussion on that ticket as to why we are waiting on the issue.

**Comment by nenaraab** [comment]
Hi @eleftherias 

thanks a lot for the hint with the `oidcLogin` Mock MVC post processor!

Honestly i did not find it, when I looked for spring test utilities... but actually i was looking for a testing approach for methods that were secured with custom `@PreAuthorize` annotations and `@WithMockUser` didn't worked.

So, in my simple demo app i was able to leverage your suggested aproach, e.g.:

```java
@Test
//@WithMockOidcUser(name = "Alice_salesOrders@test.com", authorities = {"read:salesOrders"})
public void readWith_Alice_salesOrders_200() throws Exception {
        mockMvc.perform(get("/salesOrders")
                    .with(oidcLogin()
                        .idToken(token -> token.claim("sub", "Alice_salesOrders@test.com"))
                        .authorities(new SimpleGrantedAuthority("read:salesOrders"))))
                .andExpect(status().isOk());
}
```
Not nice, but i think this becomes even nicer with an `RequestPostProcessor`... :-)

My main concern here - with this approach - is the limitation to Web MVC tests. Consider you want to test this:

```java
@Service
public class DataService {
    @PreAuthorize("hasAuthority('Admin')")
    String readSensitiveData() {
        ...
    }
}
```
This would not work efficiently with the web mvc testing utilities and everybody would have to implement its own logic to apply a valid `OidcUser` to the `SecurityContextHolder`.

So maybe you would like to reconsider this `@WithMockOidcUser` approach.

Best regards,
Nena

**Comment by eleftherias** [comment]
@nenaraab Absolutely, that all makes sense to me.
It was my mistake to close this issue, as I was unaware of the discussion in gh-8461.
Please disregard my comment and continue the discussion on the PR gh-8461.

**Comment by jzheaux** [comment]
@nenaraab, to make sure we're solving the right problem here, can you explain what problems you ran into with `@WithMockUser`?

I ask since - in cases where the controller or service has nothing OIDC-specific in it - `@WithMockUser` should work just fine.

**Comment by nenaraab** [comment]
Hi @jzheaux 

that's a fair question. You're right, depending on whether developers needs to provide further "userInfo" claims, the `@WithMockOidcUser` must be able to handle that as well.

Here some insights into my current project context:

The scenario (method i want to test):
```java
    @GetMapping(value = "/authenticate") // redirects to login page
    public String secured(@AuthenticationPrincipal OidcUser principal) {
        String name = principal.getGivenName();

        if (name == null) {
            name = principal.getEmail();
        }
        return "Congratulation, " + name
                + "! You just logged in successfully.";
    }
```

I tried to implement tests with these options:
1. using `@WithMockUser`
```java
    @Test
    @WithMockUser("Alice_countryCode@test.com")
    public void authenticateWithoutPermission_200() throws Exception {
        mockMvc.perform(get("/authenticate"))
                .andExpect(status().isOk());
    }
```
which result into a NullPointerException for `@AuthenticationPrincipal OidcUser principal`:
```
Caused by: java.lang.NullPointerException
	at com.sap.cloud.security.samples.BasicController.secured(BasicController.java:36)
	at com.sap.cloud.security.samples.BasicController$$FastClassBySpringCGLIB$$de505469.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:687)
	at com.sap.cloud.security.samples.BasicController$$EnhancerBySpringCGLIB$$7f310964.secured(<generated>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:190)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:138)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:105)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:879)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:793)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1040)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:943)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	... 126 more
```

2. using `OidcLoginRequestPostProcessor.oidcLogin()`
```java
    @Test
    public void authenticateWithoutPermission_200() throws Exception {
        mockMvc.perform(get("/authenticate").with(oidcLogin()))
                .andExpect(status().isOk());
    }
```
![image](https://user-images.githubusercontent.com/5861301/83284660-4c68c380-a1dd-11ea-90fb-da4a49369eb8.png)

3. using `@WithMockOidcUser`
```java
    @Test
    @WithMockOidcUser("Alice_countryCode@test.com")
    public void authenticateWithoutPermission_200() throws Exception {
        mockMvc.perform(get("/authenticate"))
                .andExpect(status().isOk());
    }
```
![image](https://user-images.githubusercontent.com/5861301/83285413-78387900-a1de-11ea-9596-845b5fd99b92.png)

Happy whitsun!

**Comment by jzheaux** [comment]
Got it, thanks for the extra insight, @nenaraab.

>  depending on whether developers needs to provide further "userInfo" claims, the @WithMockOidcUser must be able to handle that as well.

Since [the proposed PR](https://github.com/spring-projects/spring-security/pull/8461) doesn't address the use cases you've listed, maybe let's take a look together at what supporting claims in the annotation would look like.

Some things don't map well to annotations. In the case of OIDC, claims that are an epoch date or a JSON object map poorly. Both require custom annotations themselves

```java
@DateClaim(year=...

@AddressClaim(street=...
```

custom String formatting

```java
@DateClaim("2020-08-23...
```

 or a flattening of their attributes:

```java
address_street="my street", address_city="...
```

The result is a fair amount of complexity for only a little gain - that is, I don't imagine that most testers would be setting the `iat`, `exp`, or `address` claims. Though, if the annotation supports claims, I believe it'll be easier to support all claims in the spec than to try and decide on which. 

My hesitation to this point has been that the cost of supporting all claims outweighs the benefit of supporting both coding styles (annotations + fluent).

However, [your point about method security](https://github.com/spring-projects/spring-security/issues/8459#issuecomment-625810829) sheds light on the fact that it's not just about supporting two different testing styles. One cannot test only a `@Service` with an annotation that looks like:

```java
@PreAuthorize("hasAuthority('SCOPE_read') and principal.locale == 'en_US'")
```

using the `MockMvc` support. In that case, `@WithMockOidcUser(locale = "en_US")` would greatly simplify such a test.

This is also the case with `@Query` and tests that are testing just the `@Repository` layer:

```java
@Query("FROM Sales WHERE email = ?#{principal.email}")
```

Doing `@WithMockOidcUser(email="user@example.org")` would be very simple.

So, perhaps it's reasonable to reconsider my assumption about needing to support all claims or none. I wonder if there's a rubric we could follow to determine which claims to add support for.

**Comment by nenaraab** [comment]
Hi @jzheaux 

that's fine with me. Thanks.

IMHO, the absolut minimum set of configurable standard claims in context of `@WithMockOidcToken` could be:
- `email`
- `given_name` and
- `family_name`

The rest can be enhanced on demand... what do you think?

**Comment by jzheaux** [comment]
Sure, I think those could be reasonable, @nenaraab. 

Let's wait and see if someone has a concrete use case where the `MockMvc` support doesn't help, but OIDC-based annotation support would.

**Comment by ch4mpy** [comment]
> Let's wait and see if someone has a concrete use case where the `MockMvc` support doesn't help, but OIDC-based annotation support would.

@jzheaux do you mean something more than unit-testing a secured `@Component` that isn't a `@Controller` (such as a `@Service`) ?

**Comment by jzheaux** [comment]
Sorry, I wasn't clear. I agree that testing a `@Service` is a sensible idea and that a `@WithMockOidcUser` annotation would be helpful there.

What I'm trying to ascertain specifically is what is the minimum set of claims to support for a feature like this to be valuable to the community. It seems like listening for real-world use cases is a good way to figure out what that minimum set is.

Does that clarify what I'm looking for?

**Comment by ch4mpy** [comment]
The app I'm working on now is accessing `sub`, `email` and `preferred_username`.

In my lib, I chose to map all standard claims (from `org.springframework.security.oauth2.core.oidc.IdTokenClaimAccessor` and `org.springframework.security.oauth2.core.oidc.StandardClaimAccessor`). I don't think it's much of a problem to maintain.
